### ISSUE TITLE
Performance degradation and other issues when setting `matrix_product_state_max_bond_dimension`

### ISSUE THREAD

**Issue by mishmash** [Qiskit/qiskit-aer#1305]
### Informations

- **Qiskit Aer version**: source installation, `main` branch, commit 6b89d7c61e1695ae7f735977e91129d2fc029293
- **Python version**: 3.7.11
- **Operating system**: MacOS 10.15.7

### What is the current behavior?

The MPS simulator, when run in shots-sampling mode with the sample generation algorithm `mps_sample_measure_algorithm='mps_apply_measure'`, can give drastically different runtimes if one (1) explicitly specifies a bond dimension via `matrix_product_state_max_bond_dimension` in the `AerSimulator` constructor such that the circuit will be represented exactly without truncation, versus if one (2) uses the default values for `matrix_product_state_max_bond_dimension` (no limit) and `matrix_product_state_truncation_threshold` (1e-16) so as to let the code choose the appropriate bond dimension for exact representation. The problem seems to be more pronounced for states with larger bond dimension. For example, for an sca depth 2 ansatz [`RealAmplitudes(num_qubits=n, entanglement='sca', reps=2)`], I get the following runtime versus number of qubits (MacBook Pro, using 1 thread):

![image (3)](https://user-images.githubusercontent.com/1165341/128406102-e4ca756d-4cd9-4674-8933-49074f642c57.png)

All sca circuits can be represented exactly with an MPS of bond dimension \chi = 4^depth, hence the blue curve corresponds to using a backend with \chi = 4^2 = 16:

```
backend = AerSimulator(
    method='matrix_product_state',
    mps_sample_measure_algorithm='mps_apply_measure',
    matrix_product_state_max_bond_dimension=16,
    max_parallel_threads=1,
)
```
while the orange curve corresponds to
```
backend = AerSimulator(
    method='matrix_product_state',
    mps_sample_measure_algorithm='mps_apply_measure',
    max_parallel_threads=1,
)
```
At 60 qubits, the runtime difference for 1000 shots is about a factor 6. The dashed curves are reference n^2 parabolas (likely related to https://github.com/Qiskit/qiskit-aer/pull/1304).

I have checked the MPS print logs (see https://github.com/Qiskit/qiskit-aer/pull/1248), and they indicate that the bond dimensions chosen are identical in both cases; however, setting `matrix_product_state_max_bond_dimension=16` produces finite discarded values in the logs, while using the default settings does not. As a sanity check, I was able to confirm using snapshots that both configurations do produce the same state with perfect overlap to machine precision with the exact statevector for up to 20 qubits.

### Steps to reproduce the problem

The following code was used to produce the above graph:

```
import numpy as np
import pandas as pd
from collections import OrderedDict

import matplotlib.pyplot as plt

from qiskit.circuit.library import RealAmplitudes

from qiskit.providers.aer import AerSimulator
from qiskit import execute

np.random.seed(7)


def time_circuits(ansaetze_to_params, repeats=1):

    df = pd.DataFrame(
        columns=['runtime'],
        index=pd.MultiIndex.from_tuples(
            [(ansatz, n, chi, shots, threads, r)
             for ansatz,params in ansaetze_to_params.items()
             for n in params['ns']
             for chi in params['chis']
             for shots in params['shots']
             for threads in params['threads']
             for r in range(repeats)
            ],
            names=['ansatz', 'n', 'chi', 'shots', 'threads', 'repeat'],
        ),
        dtype='float',
    )

    for ansatz, n, chi, shots, threads, repeat in df.index:
        print(f"Working on {(ansatz, n, chi, shots, threads, repeat)} ...")

        var_form = RealAmplitudes(num_qubits=n, entanglement=ansatz[0], reps=ansatz[1])

        circ = var_form.assign_parameters(
            2*np.pi*np.random.random_sample(len(var_form.ordered_parameters)),
        )
        circ.measure_all()

        if chi > 0:
            backend = AerSimulator(
                method='matrix_product_state',
                mps_sample_measure_algorithm='mps_apply_measure',
                matrix_product_state_max_bond_dimension=int(chi),
                max_parallel_threads=threads,
            )
        elif chi == 0:  # Use default (unlimited) bond dimension
            backend = AerSimulator(
                method='matrix_product_state',
                mps_sample_measure_algorithm='mps_apply_measure',
                max_parallel_threads=threads,
            )
        elif chi == -1:
            backend = AerSimulator(
                method='statevector',
                max_parallel_threads=threads,
            )

        result = execute(circ, backend, shots=shots).result()
        df.loc[(ansatz, n, chi, shots, threads, repeat)] = float(result.time_taken)

    return df


ansaetze_to_params = OrderedDict({
    ('sca', 2): {
        'chis': [16, 0],
        'ns': [15, 30, 45, 60],
        'shots': [100, 1000],
        'threads': [1],
    },
})

df = time_circuits(ansaetze_to_params)


ansatz = ('sca', 2)
shots = 1000
threads = 1

ns_ = np.arange(15, 60+1)

fig, ax = plt.subplots(figsize=(12,10))

for chi in df.loc[ansatz].index.get_level_values('chi').unique():
#     if chi > 0:
        ax.plot(
            df.loc[ansatz].loc[:, chi, shots, threads, 0].index.get_level_values('n'),
            df.loc[ansatz].loc[:, chi, shots, threads, 0]['runtime'],
            'o-', label=f"$\chi={chi if chi > 0 else None}$",
        )
ax.plot(ns_, list(map(lambda x: 0.12*x**2, ns_)), '--', color='k', label=f"$\sim n^2$")
ax.plot(ns_, list(map(lambda x: 0.65*x**2, ns_)), '--', color='k')
ax.set_xlabel(r"$n$")
ax.set_ylabel(r"Runtime (seconds)")
ax.set_title(f"ansatz = {ansatz}, shots = {shots}, threads = {threads}")
ax.legend(bbox_to_anchor=(1, 1), loc='upper left', fontsize=18)

plt.show()
```
Furthermore, the print logs for the two cases, on a 10 qubit system, are as follows:

![image](https://user-images.githubusercontent.com/1165341/128405703-fbe4ab7c-3039-40f8-80ff-228b7b7a2848.png)

![image](https://user-images.githubusercontent.com/1165341/128405784-fe61113c-d21a-4d39-88b8-b1f600940b62.png)

### What is the expected behavior?

I would expect nearly identical behavior and runtime for the two configurations.

**Comment by mishmash** [comment]
Circuits with linear entanglement don't seem to suffer from the same performance discrepancies. Here is the same data for a linear ansatz with depth 6 (bond dimension = 2^6 = 64).

![image](https://user-images.githubusercontent.com/1165341/128424648-5a93ac80-83ba-4453-9f7e-8301dfa10085.png)

**Comment by merav-aharoni** [comment]
Regarding the general performance, following the performance enhancements in #1290 and in #1304, I am getting the following graph:
![From Ryan](https://user-images.githubusercontent.com/46567124/128730706-c2bd3c12-4ee7-475c-b753-9e4e97c7c9d9.png)
It does not appear to behave like n^2 any longer. 
When I compare running with `max_bond_dimension = 16` vs. no limit on bond dimension, I see the difference in performance that you pointed out. Here is the graph I get (I also added `max_bond_dimension = 100`:
![From Ryan - without n-squared](https://user-images.githubusercontent.com/46567124/128731076-1c87ea1f-dc97-4d4a-a18b-12fb4a05a52a.png)

**Comment by merav-aharoni** [comment]
I investigated this difference in performance and this is what I found: The bond dimensions are printed out after every operation in the circuit. However, they are not printed out after internal `swap` operations that the MPS adds before 2-qubit operations. There are cases where the bond dimension increased to 32 (and once I even saw 64), but eventually, after additional swaps, this was reduced back to 16. So the approximation is working correctly. 
This may actually be a good strategy - first run an exact version to determine the highest bond dimension, and then to use this for approximation.

**Comment by mishmash** [comment]
I see. That makes sense and also explains the rather dramatic difference between linear (where no swaps are needed) and sca. Great, so it seems https://github.com/Qiskit/qiskit-aer/pull/1304 and, to a lesser extent (since now the swaps will only happen once instead of every shot), this enforcing of the "correct" bond dimension will give very substantial performance gains for beyond-linear entanglement circuits.

**Comment by chriseclectic** [comment]
I believe this is now resolved in 0.9 release
